File: RegresionLineal/MinimosCuadrados.cpp
#include <iostream>
#include <vector>
#include <cmath>
using namespace std;
void regresionLineal(const vector<double>& x, const vector<double>& y, double& a0, double& a1);
void regresionLinealPotencial(const vector<double>& x, const vector<double>& y, double& a0, double& a1);
void regresionLinealExponencial(const vector<double>& x, const vector<double>& y, double& a0, double& a1);
double sumLin(const vector<double>& v);
double sumPot(const vector<double>& v);
double sumExp(const vector<double>& v);

int main() {
    vector<double> x = {1,2,3,4,5};
    vector<double> y = {0.5,1.7,3.4,5.7,8.4};

    double a0 = 0.0, a1 = 0.0;

    bool lineal = false;
    bool exponencial = true;
    bool potencial = false;
    bool decrecimiento = false;
    if(lineal){
        regresionLineal(x, y, a0, a1);
        cout << "La ecuación de la recta es: y = " << a0 << " + " << a1 << "x" << endl;
    }
    if (exponencial)
    {
        regresionLinealExponencial(x, y, a0, a1);
        cout << "La ecuación del modelo exponencial es: y = " << exp(a0) << "e^(" << a1<< "x)"<<endl;
    }
    if (potencial)
    {
        regresionLinealPotencial(x, y, a0, a1);
        cout << "La ecuación del modelo potencial es: y = "<< pow(10,a0) << "x^" << a1<<endl;
    }
    if(decrecimiento)
    {
        cout << "La ecuación de la recta es: y = "<< 1/a0 << "(x/()" << a1*(1/a0) << "+ x))"<<endl;
    }
    return 0;
}

double sumLin(const vector<double>& v) {
    double suma=0.0;
    int n = v.size();
    for (int i = 0; i < n; i++)
    {
        suma += v[i];
    }
    return suma;
}
double sumPot(const vector<double>& v) {
    double suma=0.0;
    int n = v.size();
    for (int i = 0; i < n; i++)
    {
        suma += log10(v[i]);
    }
    return suma;
}
double sumExp(const vector<double>& v) {
    double suma=0.0;
    int n = v.size();
    for (int i = 0; i < n; i++)
    {
        suma += log(v[i]);
    }
    return suma;
}


void regresionLineal(const vector<double>& x, const vector<double>& y, double& a0, double& a1) {
    int n = x.size();
    double sumX = sumLin(x);
    double sumY = sumLin(y);
    double sumXY = 0.0;
    double sumX2 = 0.0;

    for (int i = 0; i < n; ++i) {
        sumXY += log10(x[i]) * log10(y[i]);
        sumX2 += log10(x[i]) * log10(x[i]);
    }

    a1 = (n * sumXY - sumX * sumY) / (n * sumX2 - sumX * sumX);
    a0 = (sumY - a1 * sumX) / n;
}

void regresionLinealExponencial(const vector<double>& x, const vector<double>& y, double& a0, double& a1){
    int n = x.size();
    double sumX = sumLin(x);
    double sumY = sumExp(y);
    double sumXY = 0.0;
    double sumX2 = 0.0;

    for (int i = 0; i < n; ++i) {
        sumXY += x[i] * log(y[i]);
        sumX2 += x[i] * x[i];
    }

    a1 = (n * sumXY - sumX * sumY) / (n * sumX2 - sumX * sumX);
    a0 = (sumY - a1 * sumX) / n;
}

void regresionLinealPotencial(const vector<double>& x, const vector<double>& y, double& a0, double& a1) {
    int n = x.size();
    double sumX = sumPot(x);
    double sumY = sumPot(y);
    double sumXY = 0.0;
    double sumX2 = 0.0;

    for (int i = 0; i < n; ++i) {
        sumXY += log10(x[i]) * log10(y[i]);
        sumX2 += log10(x[i]) * log10(x[i]);
    }

    a1 = (n * sumXY - sumX * sumY) / (n * sumX2 - sumX * sumX);
    a0 = (sumY - a1 * sumX) / n;
}

File: RegresionLineal/minimoscuadrados.py
import numpy as np
import matplotlib.pyplot as plt

def regresion_lineal(x, y):
    n = len(x)
    sum_x = np.sum(x)
    sum_y = np.sum(y)
    sum_xy = np.sum(x * y)
    sum_x2 = np.sum(x * x)

    a1 = (n * sum_xy - sum_x * sum_y) / (n * sum_x2 - sum_x * sum_x)
    a0 = (sum_y - a1 * sum_x) / n
    return a0, a1

def regresion_lineal_exponencial(x, y):
    n = len(x)
    sum_x = np.sum(x)
    sum_y = np.sum(np.log(y))
    sum_xy = np.sum(x * np.log(y))
    sum_x2 = np.sum(x * x)

    a1 = (n * sum_xy - sum_x * sum_y) / (n * sum_x2 - sum_x * sum_x)
    a0 = (sum_y - a1 * sum_x) / n
    return a0, a1

def regresion_lineal_potencial(x, y):
    n = len(x)
    sum_x = np.sum(np.log10(x))
    sum_y = np.sum(np.log10(y))
    sum_xy = np.sum(np.log10(x) * np.log10(y))
    sum_x2 = np.sum(np.log10(x) * np.log10(x))

    a1 = (n * sum_xy - sum_x * sum_y) / (n * sum_x2 - sum_x * sum_x)
    a0 = (sum_y - a1 * sum_x) / n
    return a0, a1

def plot_regression(x, y, a0, a1, model_type):
    plt.scatter(x, y, color='blue', label='Data points')
    x_fit = np.linspace(min(x), max(x), 100)
    
    if model_type == 'lineal':
        y_fit = a0 + a1 * x_fit
        plt.plot(x_fit, y_fit, color='red', label=f'y = {a0:.2f} + {a1:.2f}x')
    elif model_type == 'exponencial':
        y_fit = np.exp(a0) * np.exp(a1 * x_fit)
        plt.plot(x_fit, y_fit, color='red', label=f'y = {np.exp(a0):.2f}e^({a1:.2f}x)')
    elif model_type == 'potencial':
        y_fit = 10**a0 * x_fit**a1
        plt.plot(x_fit, y_fit, color='red', label=f'y = {10**a0:.2f}x^{a1:.2f}')
    
    plt.xlabel('x')
    plt.ylabel('y')
    plt.legend()
    plt.title(f'Regression Model: {model_type.capitalize()}')
    plt.show()

def main():
    x = np.array([1, 2, 3, 4, 5])
    y = np.array([0.5, 1.7, 3.4, 5.7, 8.4])

    a0, a1 = 0.0, 0.0

    lineal = False
    exponencial = True
    potencial = False
    decrecimiento = False

    if lineal:
        a0, a1 = regresion_lineal(x, y)
        print(f"La ecuación de la recta es: y = {a0} + {a1}x")
        plot_regression(x, y, a0, a1, 'lineal')
    
    if exponencial:
        a0, a1 = regresion_lineal_exponencial(x, y)
        print(f"La ecuación del modelo exponencial es: y = {np.exp(a0)}e^({a1}x)")
        plot_regression(x, y, a0, a1, 'exponencial')
    
    if potencial:
        a0, a1 = regresion_lineal_potencial(x, y)
        print(f"La ecuación del modelo potencial es: y = {10**a0}x^{a1}")
        plot_regression(x, y, a0, a1, 'potencial')
    
    if decrecimiento:
        print(f"La ecuación de la recta es: y = {1/a0}(x/({a1*(1/a0)} + x))")

if __name__ == "__main__":
    main()